File: hybrid.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <cstdlib>
#include <ctime>

using namespace std;

// ---------------- MERGE FUNCTION ----------------
vector<int> mergeArrays(const vector<int>& left, const vector<int>& right) {
    vector<int> result;
    int i = 0, j = 0;

    while (i < left.size() && j < right.size()) {
        if (left[i] < right[j]) {
            result.push_back(left[i++]);
        } else {
            result.push_back(right[j++]);
        }
    }

    while (i < left.size())
        result.push_back(left[i++]);

    while (j < right.size())
        result.push_back(right[j++]);

    return result;
}

// ---------------- MERGE SORT ----------------
vector<int> mergeSort(const vector<int>& arr) {
    if (arr.size() <= 1)
        return arr;

    int mid = arr.size() / 2;
    vector<int> left(arr.begin(), arr.begin() + mid);
    vector<int> right(arr.begin() + mid, arr.end());

    left = mergeSort(left);
    right = mergeSort(right);

    return mergeArrays(left, right);
}

// ---------------- BINARY SEARCH ----------------
int binarySearch(const vector<int>& arr, int key, int low, int high) {
    while (low <= high) {
        int mid = low + (high - low) / 2;

        if (arr[mid] == key)
            return mid;
        else if (arr[mid] < key)
            low = mid + 1;
        else
            high = mid - 1;
    }
    return -1;
}

// ---------------- JUMP + BINARY SEARCH ----------------
int jumpBinarySearch(const vector<int>& arr, int key) {
    int n = arr.size();
    int step = sqrt(n);
    int prev = 0;

    while (prev < n && arr[min(step, n) - 1] < key) {
        prev = step;
        step += sqrt(n);
        if (prev >= n)
            return -1;
    }

    return binarySearch(arr, key, prev, min(step, n) - 1);
}

// ---------------- ADAPTIVE SEARCH ----------------
int adaptiveSearch(const vector<int>& arr, int key, int threshold) {
    int n = arr.size();

    if (n > threshold) {
        cout << "Dataset > 1000 → Using MOST OPTIMAL: Binary Search\n";
        return binarySearch(arr, key, 0, n - 1);
    } else {
        cout << "Dataset ≤ 1000 → Using Hybrid: Jump + Binary Search\n";
        return jumpBinarySearch(arr, key);
    }
}

// ---------------- MAIN PROGRAM ----------------
int main() {
    srand(time(0));

    int SIZE;
    const int THRESHOLD = 1000;

    cout << "Enter number of elements to generate: ";
    cin >> SIZE;

    vector<int> arr(SIZE);
    for (int i = 0; i < SIZE; i++) {
        arr[i] = rand() % 10000 + 1;
    }

    cout << "\nRandomly Generated Array:\n";
    for (int x : arr)
        cout << x << " ";
    cout << endl;

    vector<int> sortedArr = mergeSort(arr);

    cout << "\nArray After Merge Sort:\n";
    for (int x : sortedArr)
        cout << x << " ";
    cout << endl;

    int key;
    cout << "\nEnter the element to search: ";
    cin >> key;

    int index = adaptiveSearch(sortedArr, key, THRESHOLD);

    if (index != -1)
        cout << "Element " << key << " found at index " << index << endl;
    else
        cout << "Element " << key << " not found in the array\n";

    return 0;
}

File: hybrid.py
import random
import time

# ---------------- MERGE SORT ----------------
def merge_sort(arr):
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])

    return merge(left, right)

def merge(left, right):
    result = []
    i = j = 0

    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    result.extend(left[i:])
    result.extend(right[j:])
    return result


# ---------------- LINEAR SEARCH ----------------
def linear_search(arr, key):
    for i in range(len(arr)):
        if arr[i] == key:
            return i
    return -1


# ---------------- INTERPOLATION SEARCH ----------------
def interpolation_search(arr, key):
    low, high = 0, len(arr) - 1

    while low <= high and key >= arr[low] and key <= arr[high]:
        if arr[low] == arr[high]:
            break

        pos = low + int(
            ((high - low) / (arr[high] - arr[low])) * (key - arr[low])
        )

        if arr[pos] == key:
            return pos
        elif arr[pos] < key:
            low = pos + 1
        else:
            high = pos - 1

    return -1


# ---------------- INTERPOLATION + LINEAR SEARCH (HYBRID) ----------------
def interpolation_linear_search(arr, key):
    low, high = 0, len(arr) - 1

    while low <= high and key >= arr[low] and key <= arr[high]:
        if arr[low] == arr[high]:
            break

        pos = low + int(
            ((high - low) / (arr[high] - arr[low])) * (key - arr[low])
        )

        window = 5
        start = max(low, pos - window)
        end = min(high, pos + window)

        for i in range(start, end + 1):
            if arr[i] == key:
                return i

        if key < arr[start]:
            high = start - 1
        else:
            low = end + 1

    return -1


# ---------------- MAIN PROGRAM ----------------
if __name__ == "__main__":

    n = int(input("Enter number of elements to generate: "))

    arr = [random.randint(1, 10000) for _ in range(n)]

    print("\nGenerated Random Numbers:")
    print(arr)

    sorted_arr = merge_sort(arr)

    print("\nAfter Merge Sort:")
    print(sorted_arr)

    key = int(input("\nEnter the element to be searched: "))

    # -------- LINEAR SEARCH --------
    idx_linear = linear_search(sorted_arr, key)
    print("\nLinear Search:")
    if idx_linear != -1:
        print(f"Element FOUND at index {idx_linear}")
    else:
        print("Element NOT FOUND")

    # -------- INTERPOLATION SEARCH --------
    idx_interp = interpolation_search(sorted_arr, key)
    print("\nInterpolation Search:")
    if idx_interp != -1:
        print(f"Element FOUND at index {idx_interp}")
    else:
        print("Element NOT FOUND")

    # -------- HYBRID SEARCH (WITH TIME) --------
    start_time = time.perf_counter()
    idx_hybrid = interpolation_linear_search(sorted_arr, key)
    end_time = time.perf_counter()

    hybrid_time = (end_time - start_time) * 1_000_000  

    print("\nInterpolation + Linear (Hybrid) Search:")
    if idx_hybrid != -1:
        print(f"Element FOUND at index {idx_hybrid}")
    else:
        print("Element NOT FOUND")

    print(f"Time taken by the developed Hybrid Search is: {hybrid_time:.3f} microseconds")
